### PR TITLE
Adds math tests for transforms, rotations, and conversions (#103)

### PR DESCRIPTION
**Pull request by jtigue-bdai**
# Description

this PR adds tests for:

- scale_transform
- unscale_transform
- saturate
- normalize
- copysign
- convert_quat
- quat_conjugate
- quat_from_euler_xyz
- quat_from_matrix
- euler_xyz_from_quat
- matrix_from_euler
- quat_from_angle_axis
- axis_angle_from_quat
- skew_symmetric_matrix
- combine_transform
- subtract_transform
- compute_pose_error

Fixes # (issue)

## Type of change

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file
- [ ] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

### CODE REVIEW

File: source/isaaclab/test/utils/test_math.py [hunk 1]
```diff
@@ -35,6 +35,105 @@
...
+    expected_output_offset = (inputs + 0.5) / 2.5
+    torch.testing.assert_close(output_offset, expected_output_offset)
+    output_unscale_offset = math_utils.unscale_transform(output_offset, lower_offset, upper_offset)
+    torch.testing.assert_close(output_unscale_offset, inputs)
+
+
+@pytest.mark.parametrize("device", ("cpu", "cuda:0"))
+@pytest.mark.parametrize("size", ((5, 4, 3), (10, 2)))
+def test_saturate(device, size):
+    "Test saturate."
```
Comment (jtigue-bdai) [inline source/isaaclab/test/utils/test_math.py]:
```suggestion
    "Test saturate of a tensor of differed shapes and device."
```

File: source/isaaclab/test/utils/test_math.py [hunk 2]
```diff
@@ -35,6 +35,105 @@
...
+    upper_per_batch = 3.0 * torch.ones(size[1:], device=device)
+    output_per_batch = math_utils.saturate(input, lower_per_batch, upper_per_batch)
+    assert torch.all(torch.greater_equal(output_per_batch, lower_per_batch)).item()
+    assert torch.all(torch.less_equal(output_per_batch, upper_per_batch)).item()
+
+
+@pytest.mark.parametrize("device", ("cpu", "cuda:0"))
+@pytest.mark.parametrize("size", ((5, 4, 3), (10, 2)))
+def test_normalize(device, size):
+    """Test normalize."""
```
Comment (jtigue-bdai) [inline source/isaaclab/test/utils/test_math.py]:
```suggestion
    """Test normalize of a tensor along its last dimension and check the norm of that dimension is close to 1.0."""
```

File: source/isaaclab/test/utils/test_math.py [hunk 3]
```diff
@@ -257,6 +356,82 @@ def test_convention_converter(device):
     )
 
 
+@pytest.mark.parametrize("device", ("cpu", "cuda:0"))
+@pytest.mark.parametrize("size", ((10, 4), (5, 3, 4)))
+def test_convert_quat(device, size):
+    """Test convert_quat."""
```
Comment (jtigue-bdai) [inline source/isaaclab/test/utils/test_math.py]:
```suggestion
    """Test convert_quat from xyzw to wxyz and back to xyzw and verify the correct rolling of the tensor. Also check the correct exceptions are raised for bad inputs for the quaternion and the 'to'."""
```

File: source/isaaclab/test/utils/test_math.py [hunk 4]
```diff
@@ -257,6 +356,82 @@ def test_convention_converter(device):
...
+        math_utils.convert_quat(quat, to="xwyz")
+
+
+@pytest.mark.parametrize("device", ("cpu", "cuda:0"))
+def test_quat_conjugate_and_inv(device):
+    """Test quat_conjugate and quat_inv."""
+
+    quat = math_utils.random_orientation(1000, device=device)
+
+    value = math_utils.quat_inv(quat)
```
Comment (jtigue-bdai) [inline source/isaaclab/test/utils/test_math.py]:
```suggestion
    value = math_utils.quat_conjugate(quat)
```

File: source/isaaclab/test/utils/test_math.py [hunk 5]
```diff
@@ -257,6 +356,82 @@ def test_convention_converter(device):
...
+
+    with pytest.raises(ValueError):
+        math_utils.convert_quat(bad_quat)
+
+    with pytest.raises(ValueError):
+        math_utils.convert_quat(quat, to="xwyz")
+
+
+@pytest.mark.parametrize("device", ("cpu", "cuda:0"))
+def test_quat_conjugate_and_inv(device):
```
Comment (jtigue-bdai) [inline source/isaaclab/test/utils/test_math.py]:
```suggestion
def test_quat_conjugate(device):
```

File: source/isaaclab/test/utils/test_math.py [hunk 6]
```diff
@@ -257,6 +356,82 @@ def test_convention_converter(device):
...
+    """Test quat_conjugate and quat_inv."""
```
Comment (jtigue-bdai) [inline source/isaaclab/test/utils/test_math.py]:
```suggestion
    """Test quat_conjugate by checking the sign of the imaginary part changes but the magnitudes stay the same."""
```

File: source/isaaclab/test/utils/test_math.py [hunk 7]
```diff
@@ -402,6 +607,26 @@ def test_pose_inv():
     np.testing.assert_array_almost_equal(result, expected, decimal=DECIMAL_PRECISION)
 
 
+@pytest.mark.parametrize("device", ["cpu", "cuda:0"])
+def test_quat_to_and_from_angle_axis(device):
+    """Test axis_angle_from_quat quat_from_angle_axis."""
```
Comment (jtigue-bdai) [inline source/isaaclab/test/utils/test_math.py]:
```suggestion
    """Test that axis_angle_from_quat against scipy and that quat_from_angle_axis are the inverse of each other."""
```

File: source/isaaclab/test/utils/test_math.py [hunk 8]
```diff
@@ -462,6 +687,107 @@ def test_quat_box_minus_and_quat_box_plus(device):
             torch.testing.assert_close(delta_result, delta_angle, atol=1e-04, rtol=1e-04)
 
 
+@pytest.mark.parametrize("device", ["cpu", "cuda:0"])
+@pytest.mark.parametrize("t12_inputs", ["True", "False"])
+@pytest.mark.parametrize("q12_inputs", ["True", "False"])
+def test_combine_frame_transforms(device, t12_inputs, q12_inputs):
+    """Test combine_frame_transforms."""
```
Comment (jtigue-bdai) [inline source/isaaclab/test/utils/test_math.py]:
```suggestion
    """Test combine_frame_transforms such that inputs for delta translation and delta rotation can be None or specified."""
```

File: source/isaaclab/test/utils/test_math.py [hunk 9]
```diff
@@ -462,6 +687,107 @@ def test_quat_box_minus_and_quat_box_plus(device):
...
+
+    torch.testing.assert_close(expected_translation, translation_value, atol=1e-3, rtol=1e-5)
+    torch.testing.assert_close(math_utils.quat_unique(expected_quat), math_utils.quat_unique(quat_value))
+
+
+@pytest.mark.parametrize("device", ["cpu", "cuda:0"])
+@pytest.mark.parametrize("t02_inputs", ["True", "False"])
+@pytest.mark.parametrize("q02_inputs", ["True", "False"])
+def test_subtract_frame_transforms(device, t02_inputs, q02_inputs):
+    """Test combine_frame_transforms."""
```
Comment (jtigue-bdai) [inline source/isaaclab/test/utils/test_math.py]:
```suggestion
    """Test subtract_frame_transforms with specified and unspecified inputs for t02 and q02. Verify that it is the inverse operation to combine_frame_transforms."""
```

File: source/isaaclab/test/utils/test_math.py [hunk 10]
```diff
@@ -462,6 +687,107 @@ def test_quat_box_minus_and_quat_box_plus(device):
...
+    t02_compare, q02_compare = math_utils.combine_frame_transforms(t01, q01, t12_value, q12_value)
+
+    torch.testing.assert_close(t02_expected, t02_compare, atol=1e-3, rtol=1e-4)
+    torch.testing.assert_close(math_utils.quat_unique(q02_expected), math_utils.quat_unique(q02_compare))
+
+
+@pytest.mark.parametrize("device", ["cpu", "cuda:0"])
+@pytest.mark.parametrize("rot_error_type", ("quat", "axis_angle"))
+def test_compute_pose_error(device, rot_error_type):
+    """Test compute_pose_error."""
```
Comment (jtigue-bdai) [inline source/isaaclab/test/utils/test_math.py]:
```suggestion
    """Test compute_pose_error for different rot_error_type."""
```

File: source/isaaclab/test/utils/test_math.py [hunk 11]
```diff
@@ -547,15 +873,50 @@ def test_matrix_from_quat(device):
...
+        [0.0, 0.0, math.pi / 2.0],
+        [1.5708, -2.75, 0.1],
+        [0.1, math.pi, math.pi / 2],
+    ],
+)
+@pytest.mark.parametrize(
+    "convention", ("XYZ", "XZY", "YXZ", "YZX", "ZXY", "ZYX", "ZYZ", "YZY", "XYX", "XZX", "ZXZ", "YXY")
+)
+def test_matrix_from_euler(device, euler_angles, convention):
+    """Test matrix_from_euler."""
```
Comment (jtigue-bdai) [inline source/isaaclab/test/utils/test_math.py]:
```suggestion
    """Test matrix_from_euler against scipy for different permutations of the X,Y,Z euler angle conventions."""
```

**Review by ooctipus — APPROVED** [review]
This is looking really good, Thanks James!

**Comment by kellyguo11** [comment]
looks like a couple of the tests are failing in CI, is that expected?

**Comment by jtigue-bdai** [comment]
> looks like a couple of the tests are failing in CI, is that expected?

Nope not expected. Let me take a look.

**Comment by jtigue-bdai** [comment]
> > looks like a couple of the tests are failing in CI, is that expected?
> 
> Nope not expected. Let me take a look.

@kellyguo11 fixed the issue. it was the compute_pose_error changes. Once I reverted those and fixed the test it seemed to work on my machine.

**Comment by jtigue-bdai** [comment]
> > > looks like a couple of the tests are failing in CI, is that expected?
> > 
> > 
> > Nope not expected. Let me take a look.
> 
> @kellyguo11 fixed the issue. it was the compute_pose_error changes. Once I reverted those and fixed the test it seemed to work on my machine.

Tests pass except `test_tiled_camera`

**Comment by ooctipus** [comment]
Some tests in test_tiled_camera is a bit finicky because the rendered pixel are inherently stochastic in tiled camera, we fix some of it in the upcoming commits, but I think it should be fine. @kellyguo11

**Comment by kellyguo11** [comment]
awesome, latest run looks good!
